keps/sig-network/4444-service-routing-preference/README.md

```diff
@@ -0,0 +1,945 @@
+No.
+
+###### Will enabling / using this feature result in introducing new API types?
+
+No.
+
+###### Will enabling / using this feature result in any new calls to the cloud provider?
+
+No.
+
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+- Users using the new field will see an see an increase in the Service object of
+  ~10 bytes
+- Also, in situations when the status for this new field is reported, size
+  equivalent to the addition of two Condition types may be observed.
+
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+No.
+
+###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
+
+No.
+
+###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?
+
+No.
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+_This section will be completed when targeting beta to a release._
+
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
```

Comment:
At some point in the last discussion, we discovered [`pod.spec.topologySpreadConstraints`](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/), which are a way of saying things like "I want this Deployment deployed such that each zone/node/whatever has the same number of Pods", and there was [a brief discussion](https://github.com/kubernetes/enhancements/pull/3293#issuecomment-1322698198) about how we could use that for topology-aware service routing.

Reply:
+1 - definitely worth covering that option in this section. My personal opinion is that that API is both too complex for users and too limiting for implementations.

Reply:
Thanks for the suggestions. I've added a section on `Reuse Pod Topology Spread Constraints for Traffic Distribution`. Additionally, feels like rather than being an either-or situation, it's more appropriate for Pod Topology Spread Constraints to complement the `routingPreference`. Added a section on `Complementary Use of Pod Topology Spread Constraints and routingPreference` which eventually might also make sense in a user facing doc

> My personal opinion is that that API is both too complex for users and too limiting for implementations.

I share the same thoughts. The section on `Granular Routing Controls` below exactly emphasis those downsides.